### ISSUE TITLE
Failure of WCAG's Use Of Color reported in DevTools 4.8.2, where other contrast checkers report conformance

### ISSUE THREAD

**Issue by fstrr** [dequelabs/axe-core#4332]
### Product

axe Extension

### Product Version

4.8.2

### Latest Version

- [X] I have tested the issue with the latest version of the product

#### Expectation
Inline links with no non-color way to identify them need to have a contrast of at least 3:1 with their surrounding non-link text. I was testing a site and saw the "Links must be distinguishable without color" error. According to WebAIM's and TPGi's contrast checkers, the color combinations are conforming, so shouldn't have failed. The expectation is that colors are correctly compared and errors aren't erroneously reported. **Note**: after looking at this a little more, I'm now not 100% sure whether Axe DevTools is right or wrong here.

#### Actual
The product gives a contrast ratio of 2.51:1 between the link and non-link text, where other tools give a contrast of 3.06:1.

Axe DevTools seems to be converting the semi-transparent black color to `#212121`, which then causes the lower contrast ratio:
<img width="959" alt="screen grab of axe dev tools showing the error message" src="https://github.com/dequelabs/axe-core/assets/20049493/24294e43-70e7-41db-a410-e1316e37321e">

Compare that to [WebAIM's contrast checker](https://webaim.org/resources/contrastchecker/) using the actual values from the CSS:

<img width="687" alt="screen grab of WebAIM's contrast checker showing a passing level of contrast" src="https://github.com/dequelabs/axe-core/assets/20049493/d90c78cb-1324-4864-826d-3dcb3d282ae7">

#### How to Reproduce
[Here's a CodePen](https://cdpn.io/pen/debug/xxBMOgE) containing a sample paragraph with an inline link with no underline. The two colors are:

1. Text: `#000000de` (black with 0.87 alpha)
2. Link: `#255EAD`

Open that page in a browser and use dev tools to scan the page. Once the scan is complete, there will be a Link Must Be Distinguishable Without Color error. Expand the error message to see the description.

**Comment by dylanb** [comment]
@straker looks like a false positive to me

**Comment by straker** [comment]
I can't say for certain if this is a false positive or not. Need someone who knows the color spec better than I do. From our code standpoint though, when we try to get the foreground color we have to blend the foreground color into the background color to resolve the transparency. This uses the [simple alpha compositing function](https://www.w3.org/TR/compositing-1/#simplealphacompositing) which results in the color being `rgba(33,33,33,1)` or `#212121`. I noticed that the webaim color picker doesn't ask what the background color is so I'm not sure how they decide what the final color is since the background color will blend into the foreground color due to the opacity.

**Comment by straker** [comment]
I think I understand what the WebAIM's color contrast picker is doing. If you type `#212121` into the foreground color with a background color of `#ffffff`, then you get a contrast ratio of 16.1:1, which is the same contrast if you put `#000000de` into the foreground color. It appears that it's blending the foreground color into the background color to give the contrast. So if you put the foreground color of the link into the background color spot, I'm thinking it gives a blended color result for those two, which will be incorrect as it's trying to blend the text color into the link color.

In this case I believe axe-core is correct on the contrast between the two colors taking into account the blending of the background for the paragraph text.

**Comment by fstrr** [comment]
FWIW, [TPGi's Colour Contrast Analyzer](https://www.tpgi.com/color-contrast-checker/) has almost the same result at WebAIM's (although it claims a contrast of 3.07:1 instead of 3.06:1) when color codes are pasted into the tool:

<img width="483" alt="screen grab of TPGi's colour contrast tool showing a contrast of 3.07:1" src="https://github.com/dequelabs/axe-core/assets/20049493/85646dd2-17c6-4d06-808a-c12d2025ab0c">

TPGi's tool only allows alpha channels on foreground colors, so I can't swap the two around to get a different result.

I think your assumption is correct that WebAIM, and TPGi, are blending the two colors and, thus, getting it wrong. Incidentally, I did discover that [WebAIM has a separate link color checking tool](https://webaim.org/resources/linkcontrastchecker/?fcolor=212121&bcolor=FFFFFF&lcolor=255EAD) that I hadn't seen before but wish I'd known about.

I updated [my codepen](https://cdpn.io/pen/debug/xxBMOgE) to add a paragraph of blue text on a partially transparent background and, using the color contrast checker in Edge's dev tools, I get the same 2.5:1 contrast ratio that axe dev tools does:

<img width="243" alt="screen grab of Edge's color tool showing a contrast of 2.5:1" src="https://github.com/dequelabs/axe-core/assets/20049493/42fc50c7-8d6c-4edd-9e9e-039899579c22">

## Conclusions

1. When testing for Use Of Color with links, TPGi's tool shouldn't be used to paste in CSS values for semi-transparent colors as it's missing the essential third color to specify background.
2. Use WebAIM's link color checking tool and not their basic color contrast tool
3. Axe Dev Tools does seem to be doing the correct thing.
4. I don't think there is a bug here and this can be closed.
